Move FAQ from wiki to userguide (#5374)
cython/cython: docs/src/userguide/faq.rst

.. contents::

Basics
======

Do I need to rename my ``.py`` file to ``.pyx``?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: No. Cython can compile both .py and .pyx files. The difference is that the extended Cython syntax (``cdef …``) is only available in Cython .pyx files and not in Python .py files.

Author: Yeah, a lot of the answers will need updating. So far I've just done minimal edits.

----------

How do I speed up the C compilation?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: Especially with large modules, the code that Cython generates can take the C compiler quite some time to optimise. This is usually ok for production builds, but during development, this can get in the way.

It can substantially speed up the C compiler runs to disable the code optimisation, e.g. by setting the environment variable ``CFLAGS="-O0 -ggdb"`` on Linux or MacOS, which also enables full debugging symbols for better crash reports and debugger usage.  For MSVC on Windows, you can pass the option ``/Od`` to disable all optimisations.

----------

How do I reduce the size of the binary modules?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: The Python distutils build often includes debugging symbols in the extension modules.  The default for gcc is ``-g2``, for example. Disabling them (``CFLAGS=-g0`` for gcc), or setting them to the bare minimum that is required to produce stack traces on crashes (``CFLAGS=-g1`` for gcc), can visibly reduce the size of the binaries.

----------

> How well is Unicode supported?

Reviewer: I am pretty sure this chapter is no longer needed

Reviewer 2: The first paragraph still looks relevant to me. The second paragraph should probably go since `Py_UNICODE` is deprecated

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: The support for Unicode is as good as CPython's, but additionally distinguishes between  the Python ``str`` (``bytes`` in Python 2.7) and ``unicode`` (always Unicode text) string type. Note that there is no equivalent C type available for Unicode strings, but Cython can automatically convert (encode/decode) from and to encoded C/C++ strings (``char*`` / ``std::string``).

There is also native support for the ``Py_UCS4`` C integer type that represents a single unicode character. In fact, Cython will try to infer this type for single character unicode literals, and avoid the creation of a unicode string object for them if possible. This is because many operations work much more efficiently (in plain C) on ``Py_UCS4`` than on unicode objects.

See the `string tutorial<string_tutorial>`_.

How do I ...?
=============

How do I pickle cdef classes?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: See `the documentation<auto_pickle>`_.

----------

How do I use a Cython class in a C++ framework?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: See, for example, http://bitbucket.org/binet/cy-cxxfwk/src and http://groups.google.com/group/cython-users/browse_thread/thread/bc007d85b2ccc518 .

Author: This whole answer is highly outdated; I haven't updated it yet (how?)

----------

How do I help Cython find numpy header files?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: If you are seeing errors like these:

::

     error: numpy/arrayobject.h: No such file or directory
     error: numpy/ufuncobject.h: No such file or directory

You should modify your setup.py file to grab the numpy include directory as follows:

::

    import numpy
    ...
    setup(
        ...
        ext_modules = [Extension(..., include_dirs=[numpy.get_include()])]
    )

----------

How do I declare numeric or integer C types?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: In most cases, you don't need to. For types declared in ``stdint.h``, just ``cimport`` them from ``libc.stdint`` which comes with Cython, e.g.

::

    from libc.stdint cimport uint32_t, int64_t

>     cdef int64_t i = 5

Author: Converting this page to use both syntax variants is probably a larger operation best left for a follow-up.

        if not isinstance(text, unicode):
            raise ValueError("requires text input, got %s" % type(text))
        utf8_data = text.encode('UTF-8')
        c_handle_data( utf8_data, len(utf8_data) )

Note that this also accepts subtypes of the Python unicode type. Typing the "text" parameter as "unicode" will not cover this case.

The above is the right thing to do in Py3. However, some (not all, just some) module APIs may become more user friendly in Python 2.x if you additionally allow well defined byte strings. For example, it may make sense to allow plain ASCII strings in some cases, as they are often used for textual data in Python 2.x programs. This could be done as follows:

::

    from python_version cimport PY_MAJOR_VERSION

    def work_with_text_data(text):
        if isinstance(text, unicode): # most common case first
            utf8_data = text.encode('UTF-8')
        elif (PY_MAJOR_VERSION < 3) and isinstance(text, str):
            text.decode('ASCII') # trial decoding, or however you want to check for plain ASCII data
            utf8_data = text
        else:
            raise ValueError("requires text input, got %s" % type(text))

>         c_handle_data(utf8_data, len(utf8_data))

Author: Py2 only, remove?

Reviewer 3: Yeah, I'd remove this. It's for people who write new code for new APIs, and new code is rather unlikely to be written with Py2 in mind.

----------

How do I use variable args?
^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: It can't be done cleanly yet, but the code below works:

::

    cdef extern from "stdarg.h":
        ctypedef struct va_list:
            pass
        ctypedef struct fake_type:
            pass
        void va_start(va_list, void* arg)
        void* va_arg(va_list, fake_type)
        void va_end(va_list)
        fake_type int_type "int"

    cdef int foo(int n, ...):
        print "starting"
        cdef va_list args
        va_start(args, <void*>n)
        while n != 0:
            print n
            n = <int>va_arg(args, int_type)
        va_end(args)
        print "done"

    def call_foo():
        foo(1, 2, 3, 0)
        foo(1, 2, 0)

----------

How do I make a standalone binary from a Python program using cython?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: You probably want a recipe something like this:

::

>     PYVERSION=2.7

Author: Py2 - replace?

    foobar: foobar.py
        cython --embed foobar.py -o foobar.c
        $(CC) -I /usr/include/python$(PYVERSION) foobar.c -lpython$(PYVERSION) -o foobar

The magic is the --embed option, which embeds a copy of the Python interpreter main in the generated C.  You'll want to change 'foobar' to reflect the name of your script, of course, and PYVERSION as appropriate.

----------

How do I have to wrap C code that uses the restrict qualifier?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: There currently is no way of doing this directly into C code. Cython does not understand the restrict qualifier. However you can wrap your way around it.

Author: Is there an issue for this? -- I couldn't find one. `restrict` has been in C since C99, so this should IMO be added.

See the following example code:

slurp.h

> -------

Author: I think we want to remove these headers?

import os
os.path.expanduser('~')
```

----------

How do I access native Python file objects?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**:  See this small example of how to access Python file objects:

::

    #!python
    # Idiom for accessing Python files.
    # First, declare the Python macro to access files:
    cdef extern from "Python.h":
        ctypedef struct FILE
        FILE* PyFile_AsFile(object)
        void  fprintf(FILE* f, char* s, char* s)
    # Next, enter the builtin file class into the namespace:
    cdef extern from "fileobject.h":
        ctypedef class __builtin__.file [object PyFileObject]:
            pass
    # Now declare the C function that requires a file:
    cdef void c_printSomething(FILE* outFile, char* str):
        fprintf(outFile, "%s", str)
    # Now create a class or some other definition that uses the function:
    ctypedef class ExampleUsingFile:
        def printSomething(self, file outFile, char* str):
            c_printSomething(PyFile_AsFile(outFile), str)

with simple test:

::

    #!python
    import sys
    import file_example
    x = file_example.ExampleUsingFile()
    x.printSomething(sys.stdout, "hello world!\n")

> **Note**: This does no longer work in Python 3, where file objects have no representation at the C-API level.

Reviewer 2: I think there's a posix solution where you `file.fileno()` to get a file descriptor, the `fdopen` to get a `FILE*`. But I'm not sure of the details off the top of my head

    cdef object l = <object>ptr    # this increases the reference count to the list

Note that the lifetime of the object is only bound to its owned references, not to any C pointers that happen to point to it. This means that ``ptr`` in the example above becomes invalid as soon as the last reference to the object dies:

::

    py_object = [1,2,3]
    cdef PyObject* ptr = <PyObject*>py_object
    py_object = None   # last reference to list dies here

    # ptr now points to a dead object
    print(<object>ptr)   # expect a crash here!

Pointers are commonly used when passing objects through C callbacks, e.g.

::

    cdef int call_it_from_c(void* py_function, void* args):
        py_args = <tuple>args if args is not NULL else ()
        return (<object>py_function)(*py_args)

    def py_func(a,b,c):
        print(a,b,c)
        return -1

    args = [1,2,3]

    call_it_from_c(<PyObject*>py_func, <PyObject*>args)

Once again, care must be taken to keep the objects alive as long as any pointers to them are still in use.

----------

Why does ``**`` on int literals not work (as it seems to do in Pyrex)?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: It works as expected in recent versions of Cython.

Author: Seems obsolete - remove?

Reviewer 2: Or link to https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#power-operator

Reviewer 3: I'd rather remove it all together. There's no need to have FAQs that people won't even risk asking. If it's in the regular documentation, that's where people will rather look instead.

The C code generated by Cython is portable and builds in all supported Python versions. All supported CPython release series are tested regularly. New CPython versions are usually supported before they are released.

The source code that Cython compiles can use both Python 2 and Python 3 syntax, defaulting to Python 2 syntax in Cython 0.x and Python 3 syntax in Cython 3.x and later. When compiling Cython modules (.pyx files) in Python 2 mode, most Python 3 syntax features are available by default if they do not interfere with Python 2 syntax (as in Python 2.7), but the general language semantics are defined as in Python 2. When compiling Python modules (.py files), the special Cython syntax (such as the ``cdef`` keyword) is not available. For both input types, the language level can be set to Python 3 by either passing the "-3" option to the compiler, or by putting

::

    # cython: language_level=3

> at the top of the module file (within the first comment and before any code or empty lines). With Cython 3.x, compiling Python 2 code requires the option "-2" or the directive ``language_level=2``. The most visible difference is that unprefixed strings are unicode strings under Python 3 semantics, but this also makes ``print()`` a function, keeps loop variables in list comprehensions from leaking into the outer scope, etc. To get Python 3 syntax features without enforcing unicode literals, set ``language_level=3str`` or use the option ``--3str``.

Reviewer 3: ```suggestion
at the top of the module file (within the first comment and before any code or empty lines). With Cython 3.x, compiling Python 2 code requires the option "-2" or the directive ``language_level=2``. By default, with the Python 3 semantics in Cython 3.0, ``print()`` is a function, loop variables in list comprehensions do not leak into the outer scope, etc. This is equivalent to ``language_level=3str`` or the option ``--3str``.  If you instead select ``language_level=3``, then, additionally, unprefixed strings are always unicode strings. 
```

----------

Can I place the output under the BSD license, or does it have to be the python-license as well?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: You can use the output of Pyrex/Cython however you like (and license it how you like - be it BSD, public domain, GPL, all rights reserved, whatever).

Reviewer 2: Yeah, that look better

More details: The Python License is different from the GPL used for GCC, for example. GCC requires a special exception clause for its output as it is *linked* against the library part of GCC, i.e. against GPL software, which triggers the GPL restrictions.

> Pyrex doesn't do anything similar, and linking against Python is not restricted by the Python License, so the output belongs to the User, no other rights or restrictions involved.

Reviewer: ```suggestion
Pyrex/Cython doesn't do anything similar, and linking against Python is not restricted by the Python License, so the output belongs to the User, no other rights or restrictions involved.
```

Author: I think most of the pyrex stuff can be removed, TBH (perhaps except in the "About the project" section for historical interest).

> Pyrex doesn't do anything similar, and linking against Python is not restricted by the Python License, so the output belongs to the User, no other rights or restrictions involved.

Reviewer 3: ```suggestion
Cython doesn't do anything similar, and linking against Python is not restricted by the Python License, so the output belongs to the User, no other rights or restrictions involved.
```

in a more formal way, such as

::

    R. Bradshaw, S. Behnel, D. S. Seljebotn, G. Ewing, et al., The Cython compiler, http://cython.org.

(the list of author names were taken from setup.py)

For a yet more formal citation, there is a [[journal paper|https://www.computer.org/csdl/magazine/cs/2011/02/mcs2011020031/13rRUx0Pqtw|journal paper]] on Cython.
If you wish to cite it, here's the Bibtex:

::

    @ARTICLE{ behnel2010cython,
        author={Behnel, S. and Bradshaw, R. and Citro, C. and Dalcin, L. and Seljebotn, D.S. and Smith, K.},
        journal={Computing in Science Engineering},
        title={Cython: The Best of Both Worlds},
        year={2011},
        month=march-april ,
        volume={13},
        number={2},
        pages={31 -39},
        keywords={Cython language;Fortran code;Python language extension;numerical loops;programming language;C language;numerical analysis;},
        doi={10.1109/MCSE.2010.118},
        ISSN={1521-9615},
    }

----------

What is the relation between Cython and Pyrex? Are the barriers between the two based on technical direction? Differing goals?

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 3: Right. For a few years now already, I keep seeing blank faces whenever I mention that Cython is based on something called Pyrex. Definitely not a FAQ.

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Author: I made a suggestion for a substantially shortened question & answer here. Not with any authority of course, so please let me know if you'd like something changed/improved.

                            cback(buf.strip())
                        else:
                            cback(buf.strip(), args)

This avoids the problems using the restrict qualifiers (Such as are needed with the functions declared in regex.h on FreeBSD [at least 7.X]) by
allowing the C compiler to handle things going from C to C, Cython's support for this even using the "const trick" doesn't seem to behave properly
(at least as of 0.12). the following commands will generate your compiled module from the above source:

::

    cython -o slurp.c slurp.pyx
    cc -shared -I/usr/include -I./ -I/usr/local/include/python2.5 -L/usr/local/lib -lpthread -lpython2.5 cslurp.c slurp.c -o slurp.so

It is also possible to use distutils by adding the file cslurp.c (or your files name) to the list of files to be compiled for the extension.

----------

How do I automatically generate Cython definition files from C (.h) or C++ (.hpp) header files ?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: Several people have created scripts to parse header files and automatically produce Cython bindings.

> ### autowrap

Reviewer 2: ```suggestion
**autowrap**
```

(### doesn't seem to be valid markdown)

autowrap automatically generates python extension modules for wrapping C++ libraries based on annotated (commented) cython pxd files. Current features include wrapping of template classes, enums, free functions and static methods as well as converters from Python data types to (many) STL containers and back. Finally, also manually written Cython code can be incorporated for wrapping code.

http://github.com/uweschmitt/autowrap

> ### python-autopxd

Reviewer 2: ```suggestion
**python-autopxd**
```

        doctest.testmod(m, *args, **kwargs)

----------

How do I work around the ``-Wno-long-double error`` when installing on OS X?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**:

This is a known issue in OS X with some Python installs. It has nothing to do with Cython, and you will run
on the same trouble **every** time you want to build an C extension
module.

This is the most sane (if not the only) way to fix it:

1) Enter Python prompt, and type this:

::

    >>> from distutils import sysconfig
    >>> sysconfig.get_makefile_filename()

That should output the full path of a 'Makefile'... Open that file
with any text editor and remove  all occurrences of '-Wno-long-double'
flag.

----------

How do I work around the "unable to find vcvarsall.bat" error when using MinGW as the compiler (on Windows)?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: This error means that Python cannot find the C++ compiler on your system. Normally, this is managed by distutils, but it may happen that it's not yet up-to-date.

For example, you may be using this in setup.py:

> ```
> from distutils.core import setup
> from distutils.extension import Extension
> ```

Reviewer 2: ```suggestion
::

    from distutils.core import setup
    from distutils.extension import Extension

```

Instead, you can try to load setuptools, which will monkey-patch distutils to find vcvarsall.bat:

> ```
> try:
>     from setuptools import setup
>     from setuptools import Extension
> except ImportError:
>     from distutils.core import setup
>     from distutils.extension import Extension
> ```

Reviewer 2: ```suggestion
::

    try:
        from setuptools import setup
        from setuptools import Extension
    except ImportError:
        from distutils.core import setup
       from distutils.extension import Extension

```

In IPython, you can just import setuptools, like this:

> ```
> # First cell:
>     import setuptools
>     %load_ext Cython
> 
> # Second cell:
>     %%cython -a
>     import cython
>     cimport cython
> 
>     cdef int alpha = 255
>     print alpha
> ```

Reviewer 2: ```suggestion
::

    # First cell:
        import setuptools
                %load_ext Cython

    # Second cell:
        %%cython -a
        import cython
        cimport cython

        cdef int alpha = 255
        print alpha
```

If this is unsuccessful, try the following workarounds.

If no python libraries are imported, define the compiler by adding the following statement:

> ```
> --compiler=mingw32
> ```

Reviewer 2: ```suggestion
::

    --compiler=mingw32
```

Therefore, the line should read:

> ```
> python pyprog.py build_ext --compiler=mingw32 --inplace
> ```

Reviewer 2: ```suggestion
::
    python pyprog.py build_ext --compiler=mingw32 --inplace
```

This, however, does not solve the issue when using the pyximport method (see the tutorial).
Alternatively, the following patch can be applied.

**NOTE: This is untested.**

Open the file pyximport/pyxbuild.py and add the four lines marked with "+" at the appropriate place.

> ```diff
> diff -r 7fbe931e5ab7 pyximport/pyxbuild.py
> --- a/pyximport/pyxbuild.py Wed Sep 16 15:50:00 2009 +0200
> +++ b/pyximport/pyxbuild.py Fri Sep 18 12:39:51 2009 -0300
> @@ -55,6 +55,11 @@
> build = dist.get_command_obj('build')
> build.build_base = pyxbuild_dir
> 
> + config_files = dist.find_config_files()
> + try: config_files.remove('setup.cfg')
> + except ValueError: pass
> + dist.parse_config_files(config_files)
> +
> try:
> ok = dist.parse_command_line()
> except DistutilsArgError:
> ```

Reviewer 2: ```suggestion
::

    diff -r 7fbe931e5ab7 pyximport/pyxbuild.py
    --- a/pyximport/pyxbuild.py Wed Sep 16 15:50:00 2009 +0200
    +++ b/pyximport/pyxbuild.py Fri Sep 18 12:39:51 2009 -0300
    @@ -55,6 +55,11 @@
    build = dist.get_command_obj('build')
    build.build_base = pyxbuild_dir

    + config_files = dist.find_config_files()
    + try: config_files.remove('setup.cfg')
    + except ValueError: pass
    + dist.parse_config_files(config_files)
    +
    try:
    ok = dist.parse_command_line()
    except DistutilsArgError:
```

Finally, if this does not work, create a file called "pydistutils.cfg" in notepad and give it
the contents:

> ```ini
> [build_ext]
> compiler=mingw32
> ```

Reviewer 2: ```suggestion
::
    [build_ext]
    compiler=mingw32
```

Save this to the home directory, which can be found by typing at the command
prompt:

> ```python
> import os
> os.path.expanduser('~')
> ```

Reviewer 2: ```suggestion
::

    import os
    os.path.expanduser('~')
```

  being updated; please open an issue or a PR if you find something to improve.

Basics
======

Do I need to rename my ``.py`` file to ``.pyx``?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: No. Cython can compile both .py and .pyx files. The difference is that the extended Cython syntax (``cdef …``) is only available in Cython .pyx files and not in Python .py files.

But you can use Cython's `pure Python mode <pure-mode>`_ to provide type declarations for the compilation, including Python's PEP-484 syntax for type hints.

For cases where no interaction with external C libraries is required, this is also the recommended way to type your code, since sticking to .py files with regular Python syntax keeps the whole range of debugging, linting, formatting, profiling etc. tools for Python code available for your software development needs, which usually cannot handle the syntax of .pyx files.

----------

Can Cython generate C code for classes?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Answer**: Yes, these classes become fully fledged Python classes.

----------

Can I call my Python code from C?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: Yes, easily. Follow the example in [Demos/callback/](https://github.com/cython/cython/tree/master/Demos/callback) in the Cython source distribution.

Author: Well, the question says "from C", not "C++", so it should be alright for now?

> **Answer**: Yes, easily. Follow the example in [Demos/callback/](https://github.com/cython/cython/tree/master/Demos/callback) in the Cython source distribution.

Reviewer 2: Yeah this was an aside, not a request for a change. Leave it as is...

To instantiate an extension type, however, the fastest way is to actually use the normal Python idiom of calling the ``__new__()`` method of a type:

.. code:: python

    cdef class ExampleClass:
        cdef int _value
        def __init__(self):
            # calling "__new__()" will not call "__init__()" !
            raise TypeError("This class cannot be instantiated from Python")

    cdef ExampleClass _factory():
        cdef ExampleClass instance = ExampleClass.__new__(ExampleClass)
        instance._value = 1
        return instance

Note that this has similar restrictions as the normal Python code: it will not call the ``__init__()`` method (which makes it quite a bit faster). Also, while all Python class members will be initialised to None, you have to take care to initialise the C members. Either the ``__cinit__()`` method or a factory function like the one above are good places to do so.

How do I implement a single class method in a Cython module?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: Cython-defined methods don't bind by default, regardless from where they are referenced. Because of this the following does not work:

Author: I'm making a suggestion to rewrite the question from the POV of `binding(True)`, but would be good if someone could double-check.

> **Answer**: Cython-defined methods don't bind by default, regardless from where they are referenced. Because of this the following does not work:

Reviewer 2: I think what you've written is OK.

**Answer**: It depends on the semantics of the string. Imagine you have this C function:

::

    cdef extern from "something.h":
        cdef int c_handle_data(char* data, int length)

For __binary data__, you can simply require byte strings at the API level, so that this will work:

::

    def work_with_binary_data(bytes binary_data):
        c_handle_data(binary_data, len(binary_data))

It will raise an error (with a message that may or may not be appropriate for your use case) if users pass other things than a byte string.

For __textual data__, however, you must handle Unicode data input. What you do with it depends on what your C function accepts. For example, if it requires UTF-8 encoded byte sequences, this might work:

::

    def work_with_text_data(text):
        if not isinstance(text, unicode):
            raise ValueError("requires text input, got %s" % type(text))
        utf8_data = text.encode('UTF-8')
        c_handle_data( utf8_data, len(utf8_data) )

Note that this also accepts subtypes of the Python unicode type. Typing the "text" parameter as "unicode" will not cover this case.

----------

How do I use variable args?
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Answer**: It can't be done cleanly yet, but the code below works:

Author: > but you can use the C `va_args` mechanism.

Presumably this refers to the example below?

> **Answer**: It can't be done cleanly yet, but the code below works:

Reviewer 2: Yeah that was my intent
